src/overlay/Peer.cpp

```diff
@@ -585,10 +646,29 @@ Peer::recvMessage(StellarMessage const& stellarMsg)
     switch (stellarMsg.type())
     {
     // group messages used during handshake, process those synchronously
-    case MessageType::HELLO:
-    case MessageType::AUTH:
+    case HELLO:
+    case AUTH:
         Peer::recvRawMessage(stellarMsg);
         return;
+    case SEND_MORE:
+    {
+        cat = "CTRL";
+        // Set this once: peers aren't allowed to switch between modes
+        // Either set the flow control to "off" or continue processing the
+        // message and potentially send the next batch of items
+        if (stellarMsg.sendMoreMessage().numMessages == 0)
+        {
+            releaseAssert(!mRemotePeerWantsFlowControl);
+            mRemotePeerWantsFlowControl = std::make_shared<bool>(false);
+            // Done with the message
+            return;
+        }
+        else if (!mRemotePeerWantsFlowControl)
+        {
+            mRemotePeerWantsFlowControl = std::make_shared<bool>(true);
+        }
+        break;
```

Comment:
I am quite not sure if this is right: it seems to me that we should update `mOutboundCapacity` right away and fall through so that `maybeShedLoadAndSendNextBatch` gets called later

Reply:
not sure I understand what you're asking. Here we continue processing the message and are guaranteed to invoke `maybeShedLoadAndSendNextBatch` in case there's something in the queue to send. If you don't do this, `maybeShedLoadAndSendNextBatch` will be invoked next time we generate a new message, and I don't think that's better

Reply:
the way the code is written now, you update capacity much later via the scheduler. So any message in the scheduler that need to send messages will see the capacity as if this message was not received.